wagtail/models/__init__.py

```diff
@@ -2136,15 +2132,21 @@ def get_url_parts(self, request=None):
         if not possible_sites:
             return None
 
+        # Thanks to the ordering applied by Site.get_site_root_paths(),
+        # the first item is ideal in the vast majority of setups.
         site_id, root_path, root_url, language_code = possible_sites[0]
 
-        site = Site.find_for_request(request)
-        if site:
-            for site_id, root_path, root_url, language_code in possible_sites:
-                if site_id == site.pk:
-                    break
-            else:
-                site_id, root_path, root_url, language_code = possible_sites[0]
+        unique_site_ids = {values[0] for values in possible_sites}
+        if len(unique_site_ids) > 1 and isinstance(request, HttpRequest):
```

Comment:
I'm only just realised that I haven't explained this change on this PR very well. Not only are we guarding against an unnecessary lookup, but because this the only part of method that expects `request` to be a `HttpRequest`, this line opens the door for supplying an alternative object to treat as a cache target. Such a small change, but provides an opportunity to simplify any code that currently has to fake a `HttpRequest` just to benefit from caching (quite a strange API when you think about it)

Reply:
@gasman I know the HttpRequest requirement is something you've questioned before. How do you feel about this?

Reply:
@ababic This looks good! We can't disregard the HttpRequest entirely due to the edge case of preferring the current hostname when two sites are rooted at the same place - but minimising the surface area of code that actually depends on it being an HttpRequest is very much taking it in the right direction. I'll give this a once-over to confirm that we're not breaking any existing behaviour, but I think this is probably good to go.